plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java

```diff
@@ -319,17 +326,29 @@ private ListObjectsV2Publisher listObjectsRecursively(URI dir)
         return s3AsyncClient.listObjectsV2Paginator(request);
     }
 
-    private ListenableFuture<List<DeleteObjectsResponse>> deleteObjects(String bucketName, List<String> keys)
+    private ListenableFuture<Void> deleteObjectsInBatches(String bucketName, List<String> keys)
     {
         List<List<String>> subList = Lists.partition(keys, 1000); //  deleteObjects has a limit of 1000
         stats.getDeleteObjectsEntriesCount().add(keys.size());
-        return stats.getDeleteObjects().record(Futures.allAsList(subList.stream().map(list -> {
+        return stats.getDeleteObjects().record(asVoid(Futures.allAsList(subList.stream().map(list -> {
             DeleteObjectsRequest request = DeleteObjectsRequest.builder()
                     .bucket(bucketName)
                     .delete(Delete.builder().objects(list.stream().map(key -> ObjectIdentifier.builder().key(key).build()).collect(toImmutableList())).build())
                     .build();
             return toListenableFuture(s3AsyncClient.deleteObjects(request));
-        }).collect(toImmutableList())));
+        }).collect(toImmutableList()))));
+    }
+
+    private ListenableFuture<Void> deleteObjectsOneAtATime(String bucketName, List<String> keys)
```

Comment:
We can use batched requests for S3: https://cloud.google.com/storage/docs/batch

If there are thousands of files, trying to delete them individually seems like a bad idea.

Reply:
I see this one too, need to search for GCS APIs to do so

Reply:
Technically they are not strict blockers, but I'm still weighing on them.

Reply:
With an average request latency of `~10ms` (should we expect anything lower than that on average?) it will take half a minute to delete a directory for a normal exchange (50 tasks * 50 partitions = 2500) objects. Also I'm not sure if GCP won't start throttling us heavily because of this.

> Google's API is synchronous, whereas we want to delete asynchronously

Hmm, they really don't provide a non blocking SDK? Technically it's just an HTTP based REST API. Maybe we can use a simple HTTP client (e.g.: airlift-http-client) to issue these requests?

Reply:
sync/async API is fine, we can always use a background thread to do this.

What makes me disturbed is that with the introduction of GCS client, we will need to separate credentials code. But seems that's the best thing I can do right now.